server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java

```diff
@@ -0,0 +1,269 @@
+                    cloneCount,
+                    finalizationsCount,
+                    totalShards,
+                    completeShards,
+                    deletionsCount,
+                    snapshotDeletionsCount,
+                    activeDeletionsCount,
+                    firstStartTimeMillis,
+                    new EnumMap<SnapshotsInProgress.ShardState, Integer>(
+                        shardStatesAccumulator.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().get()))
+                    )
+                )
+            );
+
+            incompleteShardSnapshotCount += totalShards - completeShards;
+        }
+
+        return new ClusterSnapshotStats(
+            snapshotsInProgressCount,
+            incompleteShardSnapshotCount,
+            deletionsInProgressCount,
+            cleanupsInProgressCount,
+            List.copyOf(perRepositoryStats)
+        );
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+
+        builder.startObject();
+
+        builder.startObject("current_counts");
+        builder.field("snapshots", snapshotsInProgressCount);
+        builder.field("shard_snapshots", incompleteShardSnapshotCount);
+        builder.field("snapshot_deletions", deletionsInProgressCount);
+        builder.field("concurrent_operations", snapshotsInProgressCount + deletionsInProgressCount);
+        // NB cleanups are not "concurrent operations", not counted here ^
+        builder.field("cleanups", cleanupsInProgressCount);
+        builder.endObject();
+
+        builder.startObject("repositories");
+        for (PerRepositoryStats perRepositoryStats : statsByRepository) {
+            perRepositoryStats.toXContent(builder, params);
+        }
+        builder.endObject();
+
+        return builder.endObject();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(snapshotsInProgressCount);
+        out.writeVLong(incompleteShardSnapshotCount);
+        out.writeVInt(deletionsInProgressCount);
+        out.writeVInt(cleanupsInProgressCount);
+        out.writeList(statsByRepository);
+    }
+
+    public static ClusterSnapshotStats readFrom(StreamInput in) throws IOException {
+        return new ClusterSnapshotStats(
+            in.readVInt(),
+            in.readVInt(),
+            in.readVInt(),
+            in.readVInt(),
+            in.readList(PerRepositoryStats::readFrom)
+        );
+    }
+
+    record PerRepositoryStats(
+        String repositoryName,
+        String repositoryType,
+        int snapshotCount,
+        int cloneCount,
+        int finalizationsCount,
+        int totalShards,
+        int completeShards,
+        int deletionsCount,
+        int snapshotDeletionsCount,
+        int activeDeletionsCount,
+        long firstStartTimeMillis,
+        EnumMap<SnapshotsInProgress.ShardState, Integer> shardStates
```

Comment:
I wonder if this should be a `Map<String, String>` instead.

Writing and reading to stream treat keys as strings anyways.
Building json could be simplified by using `XContentBuilder#field(String name, Map<String, Object> values)`.

Map is copied anyways when creating PerRepositoryStats record.

Reply:
The main reason I prefer `EnumMap` here is that it orders the keys according to the underlying `enum`. We could achieve the same thing with `LinkedHashMap` everywhere instead but this is easy to get wrong.